code/State/SubsiteState.php

```diff
@@ -37,6 +43,36 @@ public function setSubsiteId($id)
     {
         $this->subsiteId = (int) $id;
 
+        // Persist to session, if they are enabled
+        if ($this->getUseSessions() && Injector::inst()->has(HTTPRequest::class)) {
+            Injector::inst()
+                ->get(HTTPRequest::class)
+                ->getSession()
+                ->set('SubsiteID', $id);
+        }
```

Comment:
There is a problem with this in that the session is not nested with using `->withState()` callbacks, so by setting the subsite ID in this way it would remain in the session afterwards.

Could we potentially nest the Injector in withState to avoid this?

Reply:
You should put this in the middleware then

eg..

```
class Middleware {
	function process($request, $next) {
		$state = SubsiteState::singleton();
   		$state->setSubsite($request->getSession()->get('SubsiteID'));
		try {
			return $next();
		} finally {
			$request->getSession()->set('SubsiteID', $state->getSubsiteID());
		}
}
```

That way 100% of your session access is in middleware.

Reply:
Cool, updated